Sources/Datadog/RUM/RUMVitals/VitalInfoSampler.swift

```diff
@@ -0,0 +1,92 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+import class UIKit.UIScreen
+
+internal protocol SamplingBasedVitalReader {
+    func readVitalData() -> Double?
+}
+
+internal protocol ContinuousVitalReader {
+    func register(_ valuePublisher: VitalPublisher)
+    func unregister(_ valuePublisher: VitalPublisher)
+}
+
+internal final class VitalInfoSampler {
+    private static let frequency: TimeInterval = 1.0
+
+    let cpuReader: SamplingBasedVitalReader
+    private let cpuPublisher = VitalPublisher(initialValue: VitalInfo())
+
+    var cpu: VitalInfo {
+        return cpuPublisher.currentValue
+    }
+
+    let memoryReader: SamplingBasedVitalReader
+    private let memoryPublisher = VitalPublisher(initialValue: VitalInfo())
+
+    var memory: VitalInfo {
+        return memoryPublisher.currentValue
+    }
+
+    let refreshRateReader: ContinuousVitalReader
+    private let refreshRatePublisher = VitalPublisher(initialValue: VitalInfo())
+    private let maximumRefreshRate: Double
+
+    var refreshRate: VitalInfo {
+        let info = refreshRatePublisher.currentValue
+        return info.scaledDown(by: maximumRefreshRate)
+    }
+
+    private var timer: Timer?
+
+    init(
+        cpuReader: SamplingBasedVitalReader,
+        memoryReader: SamplingBasedVitalReader,
+        refreshRateReader: ContinuousVitalReader,
+        frequency: TimeInterval = VitalInfoSampler.frequency,
+        maximumRefreshRate: Double = Double(UIScreen.main.maximumFramesPerSecond)
+    ) {
+        self.cpuReader = cpuReader
+        self.memoryReader = memoryReader
+        self.refreshRateReader = refreshRateReader
+        self.refreshRateReader.register(self.refreshRatePublisher)
+        self.maximumRefreshRate = maximumRefreshRate
+
+        takeSample()
+        let timer = Timer(
+            timeInterval: frequency,
+            target: self,
+            selector: #selector(takeSample),
+            userInfo: nil,
+            repeats: true
+        )
+        // NOTE: RUMM-1280 based on my running Example app
+        // non-main run loops don't fire the timer.
+        // Although i can't catch this in unit tests
+        RunLoop.main.add(timer, forMode: .default)
+        self.timer = timer
+    }
+
+    deinit {
+        self.timer?.invalidate()
```

Comment:
We have a `VitalInfoSampler` leak 🚰 - it's `deinit()` is never called. 

I profiled `Example` app ("tap action scenario") for a while and it has 19 instances remaining:

<img width="946" alt="Screenshot 2021-07-08 at 12 32 29" src="https://user-images.githubusercontent.com/2358722/124907646-b363c900-dfe8-11eb-8f77-a37af791dd92.png">


Here init / deinit logs for `RUMViewScope` and `VitalInfoSampler`:
```
⚡️ RUMViewScope.init() [view.name = MenuView)]
  →⚡️ VitalInfoSampler.init()
⚡️ RUMViewScope.deinit(view.name = MenuView))
⚡️ RUMViewScope.init() [view.name = TableView)]
  →⚡️ VitalInfoSampler.init()
⚡️ RUMViewScope.deinit(view.name = TableView))
⚡️ RUMViewScope.init() [view.name = MenuView)]
  →⚡️ VitalInfoSampler.init()
⚡️ RUMViewScope.deinit(view.name = MenuView))
⚡️ RUMViewScope.init() [view.name = UIControlsView)]
  →⚡️ VitalInfoSampler.init()
⚡️ RUMViewScope.deinit(view.name = UIControlsView))
⚡️ RUMViewScope.init() [view.name = MenuView)]
  →⚡️ VitalInfoSampler.init()
```

Reply:
fixed here: https://github.com/DataDog/dd-sdk-ios/pull/530#discussion_r666157508